### PR TITLE
WIP: Replacing all directory-scoped commands with target-based alternatives

### PR DESCRIPTION
**Pull request by chinz07**
## Proposed Changes

The overarching goal of this PR is to replace all directory-scoped commands with target-based alternatives.

To achieve this goal the following changes are currently implemented as part of the PR:

- Introduce imported targets for most external libraries.
- Cleanup the Boost setup code.
- Only link against the Boost components which are required by a certain target.
- Improve the distinction between private and public dependencies.

## Any background context you want to provide?

This PR is nowhere near ready to be merged. I just pushed these changes now to
run them through the CI pipeline, hopefully catching breaking changes early on.
In particular, the PR needs to be rebased onto master after #3562 has been merged.

### CODE REVIEW

File: cmake/HPX_SetupBoost.cmake [hunk 1]
```diff
@@ -23,93 +24,22 @@ set(Boost_ADDITIONAL_VERSIONS
...
-
-# If we compile natively for the MIC, we need some workarounds for certain
-# Boost headers
-# FIXME: push changes upstream
-if(HPX_PLATFORM_UC STREQUAL "XEONPHI")
-  set(Boost_INCLUDE_DIRS ${PROJECT_SOURCE_DIR}/external/asio ${Boost_INCLUDE_DIRS})
-endif()
-
 # Boost preprocessor definitions
 hpx_add_config_cond_define(BOOST_PARAMETER_MAX_ARITY 7)
```
Comment (hkaiser) [inline cmake/HPX_SetupBoost.cmake]: While this is unrelated to this PR, the `BOOST_PARAMETER_MAX_ARITY` is not needed anymore and can be removed.

**Comment by chinz07** [comment]
@hkaiser Done! Hopefully I will manage to rebase/squash the PR more frequently in the future.

File: cmake/HPX_SetupBoost.cmake (hunk 1 above)
Comment (chinz07) [reply cmake/HPX_SetupBoost.cmake]: I have removed that line to clean up the code. Thanks for the suggestion.

**Comment by hkaiser** [comment]
Some of the builds can't find the file `hpx/config/defines.hpp` anymore. This file is generated in the build directory. Did you remove some include path from the compiler's command line?

**Comment by chinz07** [comment]
@hkaiser No, the include directories are now set on a per-target basis. The build fails since some of the tests/examples do not depend on the HPX library and thus do not inherit the include directories. I just needed to figure out which targets are affected before adding the include directories explicitly.

On my machine all tests are now passing. I will push the fixes after having cleaned up the code and squash all related commits afterwards to clean up the history.

**Comment by msimberg** [comment]
@chinz07 would you mind rebasing on master again to fix the conflicts? It looks like this was actually pretty much ready to go, sorry for not following up earlier. Are you planning to still change something?

**Comment by aurianer** [comment]
Working on this :)

**Comment by msimberg** [comment]
I'm closing this since #4028 exists now.
